Add forward AD support for norm and related ops

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #74205
* #75103

Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

soulitzer committed:
Add forward AD support for norm, dist, F.pairwise_dist, F.normalize

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -14709,6 +14742,11 @@ def ref_pairwise_distance(input1, input2):
            dtypes=floating_and_complex_types_and(torch.bfloat16),
            dtypesIfCUDA=floating_and_complex_types_and(torch.float16, torch.bfloat16),
+           supports_forward_ad=True,
+           # torch.autograd.gradcheck.GradcheckError: While computing batched gradients
```
Can you open an issue for this? It looks pretty bad.

soulitzer replied:
Yep https://github.com/pytorch/pytorch/issues/74677

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -232,6 +232,72 @@ Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p
+  // NB: currently norm_jvp is also reused for dist's jvp (which haas two differentiable inputs)
+  //     but self_t still cannot be a ZT because that would require both self_t and other_t to be ZT
+  TORCH_INTERNAL_ASSERT(!self_t._is_zerotensor());
+  size_t ndim = self_p.sizes().size();  // composite compliance?
```
Use `self_p.dim()`?

soulitzer replied:
ikr

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -232,6 +232,72 @@ Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p
+    result = at::real(result);
+    return result.sum(dim, keepdim);
+  } else if (p == 2.0) {
+    auto result = self_p.sgn();
```
Why not just use `self_p` here and don't multiply it again later?

soulitzer replied:
Good catch!

albanD reviewed (commented):
If you have formulas on paper, that would be helpful to have them there to make it easier to review the code.

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -232,6 +232,72 @@ Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p
+    result = at::real(result);
+    result *= self_p.abs();
+    result = result.sum(dim, keepdim);
+    return result.div_(norm);
```
You're missing the `masked_fill_` here no? If norm has any `0` value this will go to Nan.
Is this case actually tested in OpInfo or not? IIRC we have a one off test for this.

soulitzer replied:
Yes, good catch, and also yes it should've been tested. The one-off test doesn't actually test forward AD though. I've updated the OpInfo now to test this case.

By the way, that one-off test was actually removed with the intention of subsuming it into the OpInfo, so... there was some time where it wasn't being tested for backward either. Everything should be tested now though.

soulitzer replied:
I can also see an argument for keeping one-off tests for subgradient tests: being able to see what the expected values actually are instead of having to reason about what the gradcheck is doing. The flipside is that if we have both forward AD and backward AD tests, then it starts getting more annoying to check both. So, I'm thinking as long as the test computes the subgradient as expected by the gradcheck, we can just let OpInfo handle it.

albanD replied:
Yes having OpInfo handling it is the way to go for this one.

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -232,6 +232,70 @@ Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p
+    auto result = self_p;
+    result = areAnyTensorSubclassLike({self_t}) ? result.mul(self_t.conj()) : result.mul_(self_t.conj());
+    result = at::real(result);
+    result = result.sum(dim, keepdim);
```
It will most likely be more efficient to do a `.dot()` instead of mul+sum.

soulitzer replied:
hmm still working on this one. Seems like I need to movedim, unsqueeze, and then do a matmul, squeeze, then unsqueeze the desired dims if keepdim is True.

albanD replied:
Hmmm not sure?
There might be an extra unsqueeze if you keepdim yes.
If it's more complex you can leave it like that yes.

soulitzer replied:
It is pretty complex, so I'm okay with just leaving it. I played around with it on python though and it wasn't faster on any shape, potentially due to python overhead?

```python
import torch

def movedims_to_back_and_flatten(t, dims):
  count = 0
  for dim in reversed(dims):
    count += 1
    t = t.movedim(dim, -count)
  return t.flatten(-count)

def unsqueeze_multiple(t, dims):
  for dim in dims:
    t = t.unsqueeze(dim)
  return t

def mul_then_sum(a, b, dims, keepdim):
  return (a * b).sum(dims, keepdim=keepdim)

def batched_dot(a, b, dims, keepdim):
  a_view = movedims_to_back_and_flatten(a, dims).unsqueeze(-2)
  b_view = movedims_to_back_and_flatten(b, dims).unsqueeze(-1)
  result = (a_view @ b_view).squeeze(-1).squeeze(-1)
  if keepdim:
    result = unsqueeze_multiple(result, dims)
  return result

shape = (2, 4, 5)
for keepdim in (True, False):
  for dims in ((0,), (1,), (0, 1), (0, 2), (1, 2), (0, 1, 2)):
    a = torch.rand(shape)
    b = torch.rand(shape)
    assert torch.allclose(mul_then_sum(a, b, dims, keepdim), batched_dot(a, b, dims, keepdim))
```

```
shape = (512, 256, 256)

a = torch.rand(shape, device="cpu")
b = torch.rand(shape, device="cpu")
dims = (0,)
keepdim = True
```

```
%%timeit

mul_then_sum(a, b, dims, keepdim) # 10 loops, best of 5: 82.6 ms per loop
```

```
%%timeit

batched_dot(a, b, dims, keepdim) # 1 loop, best of 5: 1.41 s per loop

```

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -232,6 +232,70 @@ Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p
+    result = result.sum(dim, keepdim);
+    return result.div_(norm);
+  } else if (std::isinf(p)) {
+    if (!keepdim && self_p.dim() != 0) {
```
Why is this only for this case?

albanD replied:
Ok!

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -232,6 +232,70 @@ Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p
+    }
+    return (at::real(self_p.sgn() * self_t.conj()) * is_eq_max / nb_max).sum(dim, keepdim);
+  } else if (p < 2.0) {
+    // TODO: This codepath can be modified to do more things inplace
```
These TODO are not necessary I think?

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

albanD reviewed (approved):
SGTM

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for norm and related ops"


Maybe things to fix:
- Not too certain about the dtype flag, but no errors were encountered while testing locally
- Forward-over-reverse grad is failing for complex but forward AD succeeds for complex and forward over reverse succeeds for real inputs
- Batched forward grad is failing due to `torch.autograd.gradcheck.GradcheckError: While computing batched gradients got: Could not allocate memory to change Tensor SizesAndStrides!`

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge this
